core/core.go

```diff
@@ -59,96 +60,84 @@ func NewCore(db ethdb.Database, config *Config, isLocalBlock func(block *types.H
-			// referencable by this block. When this block is referenced in the dom's
-			// subordinate block manifest, then ETXs produced by this block and the rollup
-			// of ETXs produced by subordinate chain(s) will become referencable.
-			if nodeCtx > common.PRIME_CTX {
-				if err := c.SendPendingEtxsToDom(types.PendingEtxs{block.Header(), newPendingEtxs}); err != nil {
-					log.Error("failed to send ETXs to domclient", "block: ", block.Hash(), "err", err)
+			if err == nil {
+				// If we have a dom, send the dom any pending ETXs which will become
+				// referencable by this block. When this block is referenced in the dom's
+				// subordinate block manifest, then ETXs produced by this block and the rollup
+				// of ETXs produced by subordinate chain(s) will become referencable.
+				if nodeCtx > common.PRIME_CTX {
+					if err := c.SendPendingEtxsToDom(types.PendingEtxs{block.Header(), newPendingEtxs}); err != nil {
+						log.Error("failed to send ETXs to domclient", "block: ", block.Hash(), "err", err)
+					}
 				}
+			} else if err.Error() == consensus.ErrFutureBlock.Error() ||
+				err.Error() == ErrBodyNotFound.Error() ||
+				err.Error() == ErrPendingEtxNotFound.Error() ||
+				err.Error() == consensus.ErrPrunedAncestor.Error() ||
+				err.Error() == consensus.ErrUnknownAncestor.Error() ||
+				err.Error() == ErrSubNotSyncedToDom.Error() ||
+				err.Error() == ErrDomClientNotUp.Error() {
+				c.addfutureHeader(block.Header())
+				log.Info("block added to future cache", "hash", block.Hash())
+			} else if err.Error() != ErrKnownBlock.Error() {
+				log.Info("error appending block", "hash", block.Hash(), "err", err)
 			}
-		} else {
-			domWait = true
 		}
 	}
 	return len(blocks), nil
 }
 
 // procfutureHeaders sorts the future block cache and attempts to append
 func (c *Core) procfutureHeaders() {
-	headers := make([]*types.Header, 0, c.futureHeaders.Len())
+	// Reconstruct future headers into future blocks list to append
+	blocks := make([]*types.Block, 0, c.futureHeaders.Len())
 	for _, hash := range c.futureHeaders.Keys() {
 		if header, exist := c.futureHeaders.Peek(hash); exist {
-			headers = append(headers, header.(*types.Header))
-		}
-	}
-	if len(headers) > 0 {
-		sort.Slice(headers, func(i, j int) bool {
-			return headers[i].NumberU64() < headers[j].NumberU64()
-		})
-
-		for _, head := range headers {
-			block, err := c.sl.ConstructLocalBlock(head)
-			if err != nil {
-				if err.Error() == ErrBodyNotFound.Error() {
-					c.addfutureHeader(head)
-				}
-				log.Debug("could not construct block from future header", "err:", err)
+			header := header.(*types.Header)
+			if block, err := c.sl.ConstructLocalBlock(header); err == nil {
+				blocks = append(blocks, block)
+				c.removeFutureHeader(header)
+			} else if err.Error() != ErrBodyNotFound.Error() &&
+				err.Error() != consensus.ErrUnknownAncestor.Error() { // Don't remove the header if we are still waiting for body
```

Comment:
I don't this this unknown ancestor check is necessary because every block in futureheader will a non dom block, and when fetcher gives a block, it has the body.

Reply:
I added this after the fact, because I was seeing this error. I can try removing it and see if the error resurfaces, but for now I don't think it hurts to have it, right?